File: forms/substance-abuse-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/substance_grader_test.rs
use substance_abuse_assessment_tera_crate::engine::substance_grader::{calculate_substance_grade, run_rules};
use substance_abuse_assessment_tera_crate::engine::substance_rules::all_rules;
use substance_abuse_assessment_tera_crate::engine::types::*;

fn empty_case() -> AssessmentData {
    AssessmentData::default()
}

#[test]
fn empty_case_is_low_risk() {
    let data = empty_case();
    let result = calculate_substance_grade(&data);
    assert_eq!(result.overall_risk, "low");
    assert_eq!(result.audit_score, 0);
    assert_eq!(result.dast_score, 0);
    assert_eq!(result.audit_risk_category, "low-risk");
    assert_eq!(result.dast_risk_category, "no-problems");
}

#[test]
fn audit_dependence_threshold_is_critical() {
    let mut data = empty_case();
    // AUDIT total = 20 (dependence likely).
    data.alcohol_use_audit.audit_q1_frequency = 4;
    data.alcohol_use_audit.audit_q2_typical_quantity = 4;
    data.alcohol_use_audit.audit_q3_binge_frequency = 4;
    data.alcohol_use_audit.audit_q4_impaired_control = 4;
    data.alcohol_use_audit.audit_q5_failed_expectations = 4;
    let result = calculate_substance_grade(&data);
    assert_eq!(result.audit_score, 20);
    assert_eq!(result.audit_risk_category, "dependence-likely");
    assert_eq!(result.overall_risk, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "AUDIT-004"));
    assert!(
        result
            .additional_flags
            .iter()
            .any(|f| f.id == "FLAG-AUDIT-001" && f.priority == "high")
    );
}

#[test]
fn dast_q3_is_inversely_scored() {
    let mut data = empty_case();
    // "no" to Q3 (able to stop) adds 1 point.
    data.drug_use_dast.dast_q3_able_to_stop = "no".to_string();
    let result = calculate_substance_grade(&data);
    assert_eq!(result.dast_score, 1);
    assert_eq!(result.dast_risk_category, "low");
}

#[test]
fn current_suicidal_ideation_fires_mh001_and_flag() {
    let mut data = empty_case();
    data.mental_health_comorbidities.suicidal_ideation_current = "yes".to_string();
    let result = calculate_substance_grade(&data);
    assert!(result.fired_rules.iter().any(|r| r.id == "MH-001"));
    assert!(
        result
            .additional_flags
            .iter()
            .any(|f| f.id == "FLAG-SI-001" && f.priority == "high")
    );
    assert_eq!(result.overall_risk, "critical");
}

#[test]
fn iv_drug_use_is_grade_3() {
    let mut data = empty_case();
    data.substance_use_history.iv_drug_use = "yes".to_string();
    let fired = run_rules(&data);
    let suh001 = fired.iter().find(|r| r.id == "SUH-001").expect("SUH-001 must fire");
    assert_eq!(suh001.grade, 3);
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), before, "rule IDs must be unique");
}

#[test]
fn liver_cirrhosis_requires_both_fields() {
    let mut data = empty_case();
    data.physical_health_impact.liver_disease = "yes".to_string();
    // Without cirrhosis type, PH-002 should not fire.
    assert!(!run_rules(&data).iter().any(|r| r.id == "PH-002"));
    data.physical_health_impact.liver_disease_type = "cirrhosis".to_string();
    assert!(run_rules(&data).iter().any(|r| r.id == "PH-002"));
}

File: forms/substance-abuse-assessment/front-end-with-svelte/src/lib/engine/substance-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateSubstanceGrade } from './substance-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { calculateAuditScore, calculateDastScore, auditRiskCategory, dastRiskCategory } from './utils';
import type { AssessmentData } from './types';

/** A blank assessment with all fields at their unanswered defaults (no store dependency). */
function blank(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', weight: null, height: null, bmi: null },
		alcoholUseAudit: {
			auditQ1Frequency: 0, auditQ2TypicalQuantity: 0, auditQ3BingeFrequency: 0, auditQ4ImpairedControl: 0,
			auditQ5FailedExpectations: 0, auditQ6MorningDrinking: 0, auditQ7Guilt: 0, auditQ8Blackout: 0,
			auditQ9Injury: 0, auditQ10Concern: 0
		},
		drugUseDast: {
			dastQ1NonMedicalUse: '', dastQ2PolyDrug: '', dastQ3AbleToStop: '', dastQ4Blackouts: '', dastQ5Guilt: '',
			dastQ6Complaints: '', dastQ7Neglect: '', dastQ8IllegalActivities: '', dastQ9Withdrawal: '', dastQ10MedicalProblems: ''
		},
		substanceUseHistory: {
			ageFirstAlcoholUse: null, ageFirstDrugUse: null, primarySubstance: '', primarySubstanceOther: '',
			secondarySubstances: '', routeOfAdministration: '', frequencyOfUse: '', durationOfUse: '', lastUseDate: '',
			currentUseStatus: '', ivDrugUse: '', needleSharing: ''
		},
		withdrawalAssessment: {
			currentlyInWithdrawal: '', withdrawalSubstance: '', tremor: '', sweating: '', nauseaVomiting: '',
			anxiety: '', agitation: '', seizureHistory: '', deliriumTremensHistory: '', hallucinations: '',
			lastDrinkDrugHours: null, withdrawalSeverity: '', medicallySupervisedDetoxNeeded: ''
		},
		mentalHealthComorbidities: {
			depression: '', depressionSeverity: '', anxietyDisorder: '', anxietyDisorderType: '', ptsd: '', ptsdDetails: '',
			bipolarDisorder: '', psychosis: '', personalityDisorder: '', eatingDisorder: '', adhd: '', suicidalIdeation: '',
			suicidalIdeationCurrent: '', selfHarmHistory: '', previousSuicideAttempts: '', psychiatricMedication: '',
			psychiatricMedicationDetails: ''
		},
		physicalHealthImpact: {
			liverDisease: '', liverDiseaseType: '', hepatitisB: '', hepatitisC: '', hivStatus: '', cardiovascularIssues: '',
			cardiovascularDetails: '', respiratoryIssues: '', respiratoryDetails: '', gastrointestinalIssues: '',
			gastrointestinalDetails: '', neurologicalIssues: '', neurologicalDetails: '', nutritionalDeficiency: '',
			chronicPain: '', chronicPainDetails: '', overdoseHistory: '', overdoseCount: null, lastOverdoseDate: ''
		},
		socialLegalImpact: {
			employmentStatus: '', occupation: '', employmentAffected: '', housingStatus: '', relationshipStatus: '',
			relationshipImpact: 'none', dependents: null, childrenSafeguardingConcerns: '', socialSupport: '',
			criminalRecord: '', criminalRecordDetails: '', currentLegalIssues: '', currentLegalDetails: '',
			duiDwiHistory: '', financialDifficulties: '', domesticViolence: '', domesticViolenceDetails: ''
		},
		previousTreatmentHistory: {
			previousTreatment: '', numberOfTreatmentEpisodes: null, previousDetox: '', detoxSetting: '', previousRehab: '',
			rehabType: '', previousCounselling: '', counsellingType: '', previousMedicationAssisted: '', matMedication: '',
			selfHelpGroups: '', selfHelpGroupType: '', longestPeriodAbstinent: '', relapseTriggers: ''
		},
		treatmentPlanningGoals: {
			treatmentGoal: '', readinessToChange: '', motivationLevel: '', preferredTreatmentSetting: '',
			interestedInCounselling: '', interestedInMedication: '', interestedInSelfHelp: '', barriersToTreatment: '',
			supportNetworkAvailable: '', supportNetworkDetails: '', riskOfRelapse: '', safetyPlanNeeded: '',
			naloxoneProvided: '', followUpPlan: ''
		}
	};
}

describe('AUDIT scoring', () => {
	it('sums the ten AUDIT items', () => {
		const d = blank();
		d.alcoholUseAudit.auditQ1Frequency = 4;
		d.alcoholUseAudit.auditQ2TypicalQuantity = 4;
		d.alcoholUseAudit.auditQ3BingeFrequency = 4;
		expect(calculateAuditScore(d.alcoholUseAudit)).toBe(12);
	});

	it('maps scores to risk categories', () => {
		expect(auditRiskCategory(5)).toBe('low-risk');
		expect(auditRiskCategory(10)).toBe('hazardous');
		expect(auditRiskCategory(17)).toBe('harmful');
		expect(auditRiskCategory(25)).toBe('dependence-likely');
	});
});

describe('DAST-10 scoring', () => {
	it('scores yes answers and the inverse Q3', () => {
		const d = blank();
		d.drugUseDast.dastQ1NonMedicalUse = 'yes';
		d.drugUseDast.dastQ2PolyDrug = 'yes';
		d.drugUseDast.dastQ3AbleToStop = 'no'; // inverse: scores
		expect(calculateDastScore(d.drugUseDast)).toBe(3);
	});

	it('maps scores to risk categories', () => {
		expect(dastRiskCategory(0)).toBe('no-problems');
		expect(dastRiskCategory(2)).toBe('low');
		expect(dastRiskCategory(4)).toBe('moderate');
		expect(dastRiskCategory(7)).toBe('substantial');
		expect(dastRiskCategory(10)).toBe('severe');
	});
});

describe('calculateSubstanceGrade', () => {
	it('returns low risk for a blank assessment', () => {
		const result = calculateSubstanceGrade(blank());
		expect(result.overallRisk).toBe('low');
		expect(result.auditScore).toBe(0);
		expect(result.dastScore).toBe(0);
	});

	it('returns critical risk for active withdrawal', () => {
		const d = blank();
		d.withdrawalAssessment.currentlyInWithdrawal = 'yes';
		const result = calculateSubstanceGrade(d);
		expect(result.overallRisk).toBe('critical');
		expect(result.firedRules.some((r) => r.id === 'WD-001')).toBe(true);
	});

	it('returns critical risk for a high AUDIT score', () => {
		const d = blank();
		d.alcoholUseAudit.auditQ1Frequency = 4;
		d.alcoholUseAudit.auditQ2TypicalQuantity = 4;
		d.alcoholUseAudit.auditQ3BingeFrequency = 4;
		d.alcoholUseAudit.auditQ4ImpairedControl = 4;
		d.alcoholUseAudit.auditQ5FailedExpectations = 4;
		const result = calculateSubstanceGrade(d);
		expect(result.auditScore).toBeGreaterThanOrEqual(20);
		expect(result.overallRisk).toBe('critical');
	});
});

describe('detectAdditionalFlags', () => {
	it('flags current suicidal ideation as high priority', () => {
		const d = blank();
		d.mentalHealthComorbidities.suicidalIdeationCurrent = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-SI-001' && f.priority === 'high')).toBe(true);
	});

	it('flags needle sharing for BBV screening', () => {
		const d = blank();
		d.substanceUseHistory.needleSharing = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-BBV-001')).toBe(true);
	});

	it('returns no flags for a blank assessment', () => {
		expect(detectAdditionalFlags(blank())).toHaveLength(0);
	});
});
